astrolabe/validator.py

```diff
@@ -106,194 +93,395 @@ def run_test_expecting_error(self, driver_workload):
+            self.fail("WorkloadExecutorError: %s" % exc)
 
         return stats
 
+    # Assert that all keys exist on the object
+    def assert_has_keys(self, obj, keys):
+        for key in keys:
+            if key not in obj:
+                self.fail("Object did not contain expected key: %s" % key)
+
+    # Assert that the parsed stats contain required fields and that numErrors
+    # and numFailures are both set (i.e. not -1)
+    def assert_basic_stats(self, stats):
+        self.assert_has_keys(stats, ['numErrors', 'numFailures', 'numIterations', 'numSuccesses'])
+
+        if stats['numErrors'] < 0:
+            self.fail_stats(
+                "Expected numErrors to be non-negative, but got {} instead."
+                .format(stats['numErrors']))
+
+        if stats['numFailures'] < 0:
+            self.fail_stats(
+                "Expected numFailures to be non-negative, but got {} instead."
+                .format(stats['numFailures']))
+
+    # Assert contents of the events.json file. The presence and type of each
+    # required field (i.e. events, errors, failures) will always be asserted.
+    # Based on the corresponding parameter for each field, the function will
+    # then assert either that the array is empty or that it is non-empty and
+    # each document therein has an expected structure.
+    #
+    # By default, this function will assert that events.json contains some
+    # events but no errors or failures.
+    #
+    # If hasErrorsXorFailures is true, hasErrors and hasFailures will be ignored
+    # and the function will instead assert that either errors or failures are
+    # present (but not both).
+    def assert_events(self, hasEvents=True, hasErrors=False, hasFailures=False,
+                      hasErrorsXorFailures=False):
+        try:
+            events = yaml.safe_load(open('events.json').read())
+        except OSError as exc:
+            self.fail("Failed to open events.json: %s" % exc)
+        except yaml.YAMLError as e:
+            self.fail("Failed to parse events.json: %s" % e)
+
+        # Assert both presence and type of required fields in events.json
+        self.assert_has_keys(events, ['events', 'errors', 'failures'])
+
+        if not isinstance(events['events'], list):
+            self.fail("The workload executor didn't record events as an array.")
+
+        if not isinstance(events['errors'], list):
+            self.fail("The workload executor didn't record errors as an array.")
+
+        if not isinstance(events['failures'], list):
+            self.fail("The workload executor didn't record failrues as an array.")
+
+        # If hasEvents is true, assert that the array is non-empty and that each
+        # object within contains essential fields. If hasEvents is false, the
+        # array should be empty.
+        #
+        # Note: we do not assert the type of events observed since not all
+        # drivers implement CMAP and thus might only log Command events.
```

Comment:
I intend to come back to this comment, as @p-mongo convinced me to leave in the code below that asserts `events` has at least one object of the types: Connection, Pool, and Command.

While I could trivially check an environment variable from within `validator.py` and disable those checks for libmongoc and PHP (neither of which have CMAP), I feel like there's a more idiomatic approach to handle this through a command line option. Perhaps @prashantmital can chime in with some advice here.

I'm currently working around this in PHPLIB (#115) by injecting fake CMAP events into the array, so it's not urgent. We can easily punt this to a separate ticket to avoid holding up this PR.

Reply:
Created [DRIVERS-1660](https://jira.mongodb.org/browse/DRIVERS-1660) to track this.